spec/src/main/asciidoc/core/invokers.asciidoc

```diff
@@ -213,24 +213,64 @@ It is a deployment problem if an attempt to identify a looked up bean results in
 Implementations are permitted to remember the identified beans and not repeat the resolution process for each invocation of `Invoker.invoke()`.
 
 All instances of `@Dependent` looked up beans obtained during `Invoker.invoke()` are destroyed before the `invoke()` method returns or throws.
+(See <<invoker_asynchronous_methods>> for a variation of this rule for asynchronous methods.)
 The order in which the instances of `@Dependent` looked up beans are destroyed is not specified.
 
-This specification recognizes the existence of _asynchronous_ methods, where the action represented by the method does not always finish when the method returns; the _completion_ of the action is asynchronous to the method call.
-For target methods that are considered asynchronous by the CDI container, the requirement to destroy instances of `@Dependent` looked up beans is relaxed: the instances of `@Dependent` looked up beans need not be destroyed before `Invoker.invoke()` returns.
-It is recommended that the instances of `@Dependent` looked up beans are destroyed after the asynchronous action completes and before the completion is propagated to the caller of `Invoker.invoke()`; if an asynchronous target method completes synchronously or throws synchronously, it is recommended that the instances of `@Dependent` looked up beans are destroyed before `invoke()` returns or rethrows the exception.
-
-[CAUTION]
-====
-The rules for recognizing asynchronous methods are not specified.
-Applications which use invokers to call asynchronous methods are therefore not portable.
-Future versions of this specification may define an API to give greater control over the invocation of asynchronous methods.
-
-Implementations that support asynchronous methods are encouraged to document the rules they follow.
-====
-
 The order in which instances of looked up beans are obtained during `Invoker.invoke()` in not specified.
 
 If an exception is thrown when creating an instance of a looked up bean during `Invoker.invoke()`, the exception is rethrown.
 
 NOTE: Destroying an instance is not permitted to throw an exception.
 See <<contextual>> for more information.
+
+[[invoker_asynchronous_methods]]
+===== Asynchronous methods
+
+For the purpose of method invokers, a target method is _asynchronous_ if it matches an _async handler_.
+In case of an asynchronous method, the action started by the method does not necessarily finish when the method returns; the _completion_ of the action is asynchronous to the method call.
+
+[source,java]
+----
+public interface AsyncHandler<T> {
+    T transform(T original, Runnable completion);
+}
+----
+
+An _async handler_ is a service provider of `jakarta.enterprise.invoke.AsyncHandler`.
+An async handler must not declare a provider method; it must declare a provider constructor.
```

Comment:
Provider constructor?

Reply:
This is specified in `java.util.ServiceLoader`.

I'm actually not sure if we need this provision, but if we allow provider methods, the service provider doesn't have to implement `AsyncHandler` and figuring out the async type would be harder. I'm not sure if anyone actually uses provider methods; the other parts of the CDI spec ignore them completely...